api/kv_v2.go

```diff
@@ -0,0 +1,403 @@
+		"data": data,
+	}
+
+	// Add options such as check-and-set, etc.
+	// We leave this as an optional arg so that most users
+	// can just pass plain key-value secret data without
+	// having to remember to put the extra layer "data" in there.
+	options := make(map[string]interface{})
+	for _, opt := range opts {
+		k, v := opt()
+		options[k] = v
+	}
+	if len(opts) > 0 {
+		wrappedData["options"] = options
+	}
+
+	secret, err := kv.c.Logical().WriteWithContext(ctx, pathToWriteTo, wrappedData)
+	if err != nil {
+		return nil, fmt.Errorf("error writing secret to %s: %w", pathToWriteTo, err)
+	}
+	if secret == nil {
+		return nil, fmt.Errorf("no secret was written to %s", pathToWriteTo)
+	}
+
+	metadata, err := extractVersionMetadata(secret)
+	if err != nil {
+		return nil, fmt.Errorf("secret was written successfully, but unable to view version metadata from response: %w", err)
+	}
+
+	kvSecret := &KVSecret{
+		Data:     nil, // secret.Data in this case is the metadata
+		Metadata: metadata,
+		Raw:      secret,
+	}
+
+	err = kvSecret.addCustomMetadata(secret)
+	if err != nil {
+		return nil, fmt.Errorf("error reading custom metadata for secret at %s: %w", pathToWriteTo, err)
+	}
+
+	return kvSecret, nil
+}
+
+// Delete deletes the most recent version of a secret from the KV v2
+// secrets engine. To delete an older version, use DeleteVersions.
+func (kv *kvv2) Delete(ctx context.Context, secretPath string) error {
+	pathToDelete := fmt.Sprintf("%s/data/%s", kv.mountPath, secretPath)
+
+	_, err := kv.c.Logical().DeleteWithContext(ctx, pathToDelete)
+	if err != nil {
+		return fmt.Errorf("error deleting secret at %s: %w", pathToDelete, err)
+	}
+
+	return nil
+}
+
+// DeleteVersions deletes the specified versions of a secret from the KV v2
+// secrets engine. To delete the latest version of a secret, just use Delete.
+func (kv *kvv2) DeleteVersions(ctx context.Context, secretPath string, versions []int) error {
+	// verb and path are different when trying to delete past versions
+	pathToDelete := fmt.Sprintf("%s/delete/%s", kv.mountPath, secretPath)
+
+	if len(versions) > 0 {
+		var versionsToDelete []string
+		for _, version := range versions {
+			versionsToDelete = append(versionsToDelete, strconv.Itoa(version))
+		}
+		versionsMap := map[string]interface{}{
+			"versions": versionsToDelete,
+		}
+		_, err := kv.c.Logical().WriteWithContext(ctx, pathToDelete, versionsMap)
+		if err != nil {
+			return fmt.Errorf("error deleting secret at %s: %w", pathToDelete, err)
+		}
+	}
+
+	return nil
+}
+
+// addCustomMetadata adds custom metadata onto the top level of the KV Secret,
+// rather than inside the KVVersionMetadata struct like in the response from
+// `vault kv get` in the CLI, to help prevent users of KVMetadata (the
+// equivalent of the response from `vault kv metadata get` in the CLI) from
+// mistakenly thinking there is such a thing as version-specific custom_metadata.
+func (kvs *KVSecret) addCustomMetadata(secret *Secret) error {
```

Comment:
This could probably be refactored into a helper function like the other `extract`s for consistency

Reply:
The reason I had this signature looking a little different than the `extract*` ones was because for some reason I have it in my head that if I'm going to be modifying the struct I have a pointer to (the `kvs *KVSecret` here), it's more idiomatic to have it as the method receiver instead of passing it in as a parameter like I've got with `secret *Secret` (which isn't being modified). 

Is that not necessarily true? Am I getting that idea from nowhere? 😆

Reply:
I was suggesting writing it as a helper, something like:
```go
func extractCustomMetadata(secret *Secret) (map[string]interface{}, error) { ... }

cM, err := extractCustomMetadata(secret)
	if err != nil {
		return nil, fmt.Errorf("error reading custom metadata for secret at %s: %w", pathToRead, err)
	}
kvS.CustomMetadata = cM
```

Reply:
I was suggesting writing it as a helper, something like:
```go
func extractCustomMetadata(secret *Secret) (map[string]interface{}, error) { ... }

cM, err := extractCustomMetadata(secret)
	if err != nil {
		return nil, fmt.Errorf("error reading custom metadata for secret at %s: %w", pathToRead, err)
	}
kvS.CustomMetadata = cM
```